### PR TITLE
Fine tune layer selector

### PR DESCRIPTION
**Pull request by nobuti**
This PR fixes cartodb.js side of https://github.com/CartoDB/cartodb/issues/10093

In addition to the issue's description, the layer selector should only be visible in the settings view in the Builder. That's why we need to segment when the vis is rendered from an embed view (by default) or by Builder.

cc @alonsogarciapablo @xavijam

### CODE REVIEW

File: src/api/create-vis.js [hunk 1]
```diff
@@ -64,13 +64,19 @@ var loadVizJSON = function (el, visModel, vizjsonData, options) {
     showLayerSelector = vizjson.options.layer_selector;
   }
 
+  var isEmbed = true;
```
Comment (alonsogarciapablo) [inline src/api/create-vis.js]: IMO, we shouldn't use the term `embed` in the context of CartoDB.js.... I'd try to find a more generic name...

File: src/geo/ui/legends/layer-legends-view.js [hunk 2]
```diff
@@ -29,29 +31,67 @@ var LayerLegendsView = Backbone.View.extend({
...
+  _shouldLegendsBeVisible: function () {
+    var showLegends = this.settingsModel.get('showLegends');
+    return showLegends && this._isLayerVisible();
+  },
+
+  _shouldLayerLegendsBeVisible: function () {
+    var isEmbed = this._isEmbed;
+    var showLayerSelector = this.settingsModel.get('showLayerSelector');
+    var showLegends = this.settingsModel.get('showLegends');
+    var isSettingsView = !!this.settingsModel.get('settingsView');
```
Comment (alonsogarciapablo) [inline src/geo/ui/legends/layer-legends-view.js]: Same here... CartoDB.js shouldn't know anything about "settings"... Are we initialising this anyway?

File: src/geo/ui/legends/layer-legends-view.js [hunk 3]
```diff
@@ -29,29 +31,67 @@ var LayerLegendsView = Backbone.View.extend({
...
+    var hasLegends = this.model.legends.hasAnyLegend();
+    var shouldVisible;
+
+    if (!isEmbed) {
```
Comment (alonsogarciapablo) [inline src/geo/ui/legends/layer-legends-view.js]: We can simplify this a bit... eg:

```
    var areLayerLegendsVisible = this._isLayerVisible() && showLegends && hasLegends;
    if (isEmbed || isSettingsView) {
      areLayerLegendsVisible = areLayerLegendsVisible || showLayerSelector;
    }
    return areLayerLegendsVisible;
```

File: src/geo/ui/legends/layer-legends-view.js (hunk 2 above)
Comment (nobuti) [inline src/geo/ui/legends/layer-legends-view.js]: kind of, `!!undefined === false`, right?

**Comment by nobuti** [comment]
`showLayerSelector` comes from the user's settings, through `layer_selector` option, IMO is a bad idea to mess with these values, they should remain inmutable. I've changed a bit the approach but avoiding to touch the options from settings.

File: src/geo/ui/legends/layer-legends-view.js [hunk 4]
```diff
@@ -29,29 +29,57 @@ var LayerLegendsView = Backbone.View.extend({
...
 
   render: function () {
-    var showLegends = this.settingsModel.get('showLegends');
-    var showLayerSelector = this.settingsModel.get('showLayerSelector');
-    var shouldVisible = showLayerSelector || this.model.legends.hasAnyLegend() && showLegends;
+    var showLegends = this._shouldLegendsBeVisible();
+    var showLayerSelector = this._shouldLayerSelectorBeVisible();
+    var shouldVisible = this._shouldLayerLegendsBeVisible();
+
+    this.$el.html(
```
Comment (alonsogarciapablo) [inline src/geo/ui/legends/layer-legends-view.js]: I guess, we could skip this if `shouldVisible` is `false`, right?

```
if (this._isVisible()) {
  this.$el.html('');
} else {
  this.$el.html(template({ ... }));
}
```

File: src/api/create-vis.js [hunk 5]
```diff
@@ -75,7 +80,8 @@ var loadVizJSON = function (el, visModel, vizjsonData, options) {
 
   visModel.setSettings({
     showLegends: showLegends,
-    showLayerSelector: showLayerSelector
+    showLayerSelector: showLayerSelector,
+    layerSelectorEnabled: layerSelectorEnabled
```
Comment (alonsogarciapablo) [inline src/api/create-vis.js]: How will the Builder change this while on the settings page?

**Pull request merged**